### PR TITLE
Nonlinear Laplacian Handling in MOLFiniteDifference (issue #354)

### PR DESCRIPTION
**Pull request by emmanuellujan**
Fixes part of #354

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
Is there a reason why this is commented?

**Comment by emmanuellujan** [comment]
I am working with some test files in my computer that isolate the portion of code that performs this particular parsing. This code is not integrated to MOL_discretization.jl yet, first I need to solve some issues with the automatic generation of rules involving Differentials. Because I wanted to upload some progress, I added the portion of code of interest at about the place where it should be integrated, but as a comment.

**Comment by emmanuellujan** [comment]
I added code to handle cases like the following: 
- Dt(u(t,x)) ~ Dx(u(t,x)^(-1) * Dx(u(t,x)))
- Dt(u(t,x)) ~ Dx(u(t,x)^2 * Dx(u(t,x)))
- Dt(u(t,x)) ~ Dx(1. / (1. + u(t,x)^2) * Dx(u(t,x)))
- Dt(u(t,x)) ~ Dx(1. / (u(t,x)^2 - 1.) * Dx(u(t,x)))
- Dt(u(t,x,y)) ~ Dx( (u(t,x,y)^2 / exp(x+y)^2 + sin(x+y+4t)^2)^0.5 * Dx(u(t,x,y))) +
                       Dy( (u(t,x,y)^2 / exp(x+y)^2 + sin(x+y+4t)^2)^0.5 * Dy(u(t,x,y)))

To do this I partially used the SymbolicUtils `@rules`.

Also, I added new test cases for all the examples above.

The function `g`  (see issue "https://github.com/SciML/DiffEqOperators.jl/issues/354") needs to be improved. Currently a basic approximation is implemented. I would like to solve this in another PR. Now I would like to address convection and convection-diffusion.

**Comment by emmanuellujan** [comment]
@ChrisRackauckas if you agree, you can merge this PR.

**Comment by ChrisRackauckas** [comment]
It's WIP and has merge conflicts?

File: src/MOLFiniteDifference/MOL_discretization.jl
Comment (ChrisRackauckas) [inline src/MOLFiniteDifference/MOL_discretization.jl:2]: ?

File: src/MOLFiniteDifference/MOL_discretization.jl
Comment (ChrisRackauckas) [inline src/MOLFiniteDifference/MOL_discretization.jl:30]: rebase?

File: src/MOLFiniteDifference/MOL_discretization.jl
Comment (ChrisRackauckas) [inline src/MOLFiniteDifference/MOL_discretization.jl:164]: why do you need to explicitly specialize on `+`?

File: src/MOLFiniteDifference/MOL_discretization.jl
Comment (ChrisRackauckas) [inline src/MOLFiniteDifference/MOL_discretization.jl:173]: why does this need to be done manually? @YingboMa @shashi don't the rules check the variations?

**Comment by emmanuellujan** [comment]
> Is there a reason why this is commented?

I update my fork from time to time, I estimate that the conflicts were generated because of that. Unfortunately I do not see an option in github that allows me to resolve them. If there is such an option, and you could point me to it, I would be happy to resolve them.

**Comment by emmanuellujan** [comment]
> It's WIP and has merge conflicts?

I update my fork from time to time, I estimate that the conflicts were generated because of that. Unfortunately I do not see an option in github that allows me to resolve them. If there is such an option, and you could point me to it, I would be happy to resolve them.

File: src/MOLFiniteDifference/MOL_discretization.jl
Comment (emmanuellujan) [reply src/MOLFiniteDifference/MOL_discretization.jl]: Thanks for pointing this out to me, I just removed those comments.

File: src/MOLFiniteDifference/MOL_discretization.jl
Comment (emmanuellujan) [reply src/MOLFiniteDifference/MOL_discretization.jl:28]: I am not sure about the meaning of the question.

File: src/MOLFiniteDifference/MOL_discretization.jl
Comment (emmanuellujan) [reply src/MOLFiniteDifference/MOL_discretization.jl:162]: My goal is to run classical convection-diffusion equations. To do so, in this PR I addressed summations involving nonlinear Laplacians, needed in diffusive terms.

File: src/MOLFiniteDifference/MOL_discretization.jl
Comment (emmanuellujan) [reply src/MOLFiniteDifference/MOL_discretization.jl:171]: Unfortunately, I did not find a way to fully use SymbolicUtils to solve this issue. So I solved a part of it manually. An explanation of the my issue can be found in the following document.: https://docs.google.com/document/d/1qbWokDYXuO5GV3ilI8R4fWWnvkBkrpQd3H6d7l9DrC4/edit

**Comment by tinosulzer** [comment]
I think you need write access to resolve conflicts on github, but if you merge master into your local branch you will see them and be able to resolve them

**Comment by emmanuellujan** [comment]
> I think you need write access to resolve conflicts on github, but if you merge master into your local branch you will see them and be able to resolve them

Thanks for the advice. In summary this is what I did:
```
git clone https://github.com/emmanuellujan/DiffEqOperators.jl/
cd DiffEqOperators.jl
git remote add upstream https://github.com/SciML/DiffEqOperators.jl/
git fetch upstream
git checkout master
git merge upstream/master
git checkout fix-issue-354
git merge master
# Solve conflict in editor 
git add src/MOLFiniteDifference/MOL_discretization.jl
git push
```

**Comment by emmanuellujan** [comment]
@ChrisRackauckas @tinosulzer

File: src/MOLFiniteDifference/MOL_discretization.jl
Comment (tinosulzer) [reply src/MOLFiniteDifference/MOL_discretization.jl:30]: edit the PR, change the base branch from SciML:master to anything else, then edit again and change back to SciML:master. This will get rid of changes like this which are just git struggling to keep track (maybe there is a smarter way haha)

File: src/MOLFiniteDifference/MOL_discretization.jl
Comment (tinosulzer) [inline src/MOLFiniteDifference/MOL_discretization.jl:127]: hmm, weird. I don't mind taking a look at this again after it's merged.

File: src/MOLFiniteDifference/MOL_discretization.jl
Comment (tinosulzer) [inline src/MOLFiniteDifference/MOL_discretization.jl:180]: why does this need 2 separates vcats?

File: src/MOLFiniteDifference/MOL_discretization.jl
Comment (tinosulzer) [inline src/MOLFiniteDifference/MOL_discretization.jl:157]: note harmonic mean might be more accurate http://web.cecs.pdx.edu/~gerry/class/ME448/lecture/pdf/CVFDdiffusion2D.pdf

**Review by tinosulzer — APPROVED** [review]
I am ok merging this as something that works for now so we can move forward. But I am not sure this is the most accurate approach: currently we do D(u_i+1/2, x_i+1/2) (i.e. function of the average) but usually I have seen D_i+1/2 (i.e. average of the function). These can be quite different if D is nonlinear. Would need to dig more into the numerical analysis to be sure which is best

File: src/MOLFiniteDifference/MOL_discretization.jl
Comment (emmanuellujan) [reply src/MOLFiniteDifference/MOL_discretization.jl:30]: Done :)

**Comment by emmanuellujan** [comment]
> I am ok merging this as something that works for now so we can move forward. But I am not sure this is the most accurate approach: currently we do D(u_i+1/2, x_i+1/2) (i.e. function of the average) but usually I have seen D_i+1/2 (i.e. average of the function). These can be quite different if D is nonlinear. Would need to dig more into the numerical analysis to be sure which is best

Currently a basic approximation is implemented. I would like to solve this in another PR.  In the first page of this [draft](https://docs.google.com/document/d/1qbWokDYXuO5GV3ilI8R4fWWnvkBkrpQd3H6d7l9DrC4/edit) you can find the definition of g based on a degree 2 polynomial in a 3D case. If this is the correct definition, I would need to know which stencil nodes should I use. The "cross" gives me 7, but I need 10 to fit the coefficients, I think that the crossed derivatives could be avoided to use just 7 nodes.

File: src/MOLFiniteDifference/MOL_discretization.jl
Comment (emmanuellujan) [reply src/MOLFiniteDifference/MOL_discretization.jl:127]: Thanks!

File: src/MOLFiniteDifference/MOL_discretization.jl
Comment (emmanuellujan) [reply src/MOLFiniteDifference/MOL_discretization.jl:157]: Thanks for the note. I added to my [draft](https://docs.google.com/document/d/1qbWokDYXuO5GV3ilI8R4fWWnvkBkrpQd3H6d7l9DrC4/edit).

File: src/MOLFiniteDifference/MOL_discretization.jl
Comment (emmanuellujan) [reply src/MOLFiniteDifference/MOL_discretization.jl]: Thanks, I am using one vcat now :)

**Comment by ChrisRackauckas** [comment]
>Currently a basic approximation is implemented. I would like to solve this in another PR. In the first page of this draft you can find the definition of g based on a degree 2 polynomial in a 3D case. If this is the correct definition, I would need to know which stencil nodes should I use. The "cross" gives me 7, but I need 10 to fit the coefficients, I think that the crossed derivatives could be avoided to use just 7 nodes.

You split it to the 3 terms Del u = u_xx + u_yy + u_zz  and then you have 2 unique coefficients per, + 1 for the middle. Otherwise you have to have the corner terms, which you could discretize to include the corners.

File: src/MOLFiniteDifference/MOL_discretization.jl
Comment (ChrisRackauckas) [reply src/MOLFiniteDifference/MOL_discretization.jl:177]: @YingboMa can you double check this?

File: src/MOLFiniteDifference/MOL_discretization.jl
Comment (ChrisRackauckas) [inline src/MOLFiniteDifference/MOL_discretization.jl:130]: just remove

File: src/MOLFiniteDifference/MOL_discretization.jl
Comment (ChrisRackauckas) [inline src/MOLFiniteDifference/MOL_discretization.jl:132]: ? just remove if there's a new definition

Comment (tinosulzer) [reply src/MOLFiniteDifference/MOL_discretization.jl:132]: it's back to the old definition which ignores the discretization order (I think) i.e. breaks #365 but not enough to cause tests to fail

**Comment by ChrisRackauckas** [comment]
>I am ok merging this as something that works for now so we can move forward. But I am not sure this is the most accurate approach: currently we do D(u_i+1/2, x_i+1/2) (i.e. function of the average) but usually I have seen D_i+1/2 (i.e. average of the function). These can be quite different if D is nonlinear. Would need to dig more into the numerical analysis to be sure which is best

That would make the approximation O(dx) instead of O(dx^2), but that can be improved in a subsequent PR. Let's get an issue open for this.

File: src/MOLFiniteDifference/MOL_discretization.jl
Comment (ChrisRackauckas) [reply src/MOLFiniteDifference/MOL_discretization.jl:132]: weird.

Comment (ChrisRackauckas) [reply src/MOLFiniteDifference/MOL_discretization.jl:132]: Are you looking into this?

Comment (tinosulzer) [reply src/MOLFiniteDifference/MOL_discretization.jl:132]: I can take a look tomorrow

**Comment by emmanuellujan** [comment]
> > Currently a basic approximation is implemented. I would like to solve this in another PR. In the first page of this draft you can find the definition of g based on a degree 2 polynomial in a 3D case. If this is the correct definition, I would need to know which stencil nodes should I use. The "cross" gives me 7, but I need 10 to fit the coefficients, I think that the crossed derivatives could be avoided to use just 7 nodes.
> 
> You split it to the 3 terms Del u = u_xx + u_yy + u_zz and then you have 2 unique coefficients per, + 1 for the middle. Otherwise you have to have the corner terms, which you could discretize to include the corners.

I know, but if you want to fit a degree 2 polynomial in 3D you need 10 coefficients, ergo 10 nodes. If you use a typical stencil (without the corners), you have 7 nodes, ergo you need to neglect some terms of the polynomial. I think the cross derivative terms could be removed to this end. Another option is to use the corners. In this case, you have 15 nodes available, but you only need 10, so you need to chose which corners to use. Probably any choice in which the nodes are sufficiently distributed is enough.

File: src/MOLFiniteDifference/MOL_discretization.jl
Comment (tinosulzer) [reply src/MOLFiniteDifference/MOL_discretization.jl:132]: The only thing that is broken with the `stencil` approach is "Test 00: Dt(u(t,x,y)) ~ Dxx(u(t,x,y)) + Dyy(u(t,x,y))" in the `MOL_2D_Diffusion.jl` test with orders 4 and 6. The test breaks because the solution is not accurate enough, not a syntax error. The reason it works with the old syntax is that it just ignores `discretization.centered_order` and pretends you're always asking for order 2. @emmanuellujan if you only test for order 2 in that test then you can use the syntax you commented out, and we should open an issue for the order 4 and 6 in 2D.

**Comment by emmanuellujan** [comment]
> > I am ok merging this as something that works for now so we can move forward. But I am not sure this is the most accurate approach: currently we do D(u_i+1/2, x_i+1/2) (i.e. function of the average) but usually I have seen D_i+1/2 (i.e. average of the function). These can be quite different if D is nonlinear. Would need to dig more into the numerical analysis to be sure which is best
> 
> That would make the approximation O(dx) instead of O(dx^2), but that can be improved in a subsequent PR. Let's get an issue open for this.

Okay, we need to wait until this PR is merged to post the new issue.

Note: currently issue #371 contains this "sub-issue".

File: src/MOLFiniteDifference/MOL_discretization.jl
Comment (emmanuellujan) [reply src/MOLFiniteDifference/MOL_discretization.jl:132]: I understand that it is better not to remove those comments for now.

File: src/MOLFiniteDifference/MOL_discretization.jl
Comment (emmanuellujan) [reply src/MOLFiniteDifference/MOL_discretization.jl:130]: See below

File: src/MOLFiniteDifference/MOL_discretization.jl
Comment (tinosulzer) [reply src/MOLFiniteDifference/MOL_discretization.jl:132]: actually I'll just open a PR to your fork

Comment (emmanuellujan) [reply src/MOLFiniteDifference/MOL_discretization.jl:132]: Oh, we posted almost at the same time, I did not see your comment when I posted mine. I will check what you mention. Thank you :)

File: src/MOLFiniteDifference/MOL_discretization.jl
Comment (ChrisRackauckas) [inline src/MOLFiniteDifference/MOL_discretization.jl:203]: ?

Comment (ChrisRackauckas) [reply src/MOLFiniteDifference/MOL_discretization.jl:203]: what exactly is this doing?

Comment (emmanuellujan) [reply src/MOLFiniteDifference/MOL_discretization.jl:203]: This code allows the non linear laplacian rules (`rules`) to be used as an argument of the function `substitute`, in an analogous way to what is done with the central derivative rules (see the code below). The first `for` iterates over the terms of the equation. The second `for` iterates over the non-linear laplacian rules (thus, a version of the rules that can not be used as an argument of the function `substitute`). If the current rule can be applied to the current term, a new rule in the correct format is generated.
Another option is not to use the `substitute` function to apply the nonlinear laplacian rules, but if we want to be consistent with what was done with the central derivative rules, In think that we have to make a transformation. Suggestions are welcome :)

Comment (ChrisRackauckas) [reply src/MOLFiniteDifference/MOL_discretization.jl:203]: No but what is `r(t) !== nothing` actually doing, and is that public API?

Comment (emmanuellujan) [reply src/MOLFiniteDifference/MOL_discretization.jl:203]: `r(t)` is applying the rule `r` to the term `t`, if the rule does not match with the term it returns `nothing`, otherwise it returns the transformed term. When we use `@rule`  (see code above) we need to work with the rule as it were a function, therefore we cannot directly use `substitute` as it is used with the central derivative rules. 

It is in the public API ("https://symbolicutils.juliasymbolics.org/api/"):

> `rule LHS => RHS`
> Creates a `Rule` object. A rule object is callable, and takes an expression and rewrites it if it matches the LHS pattern to the RHS pattern, returns nothing otherwise.

Comment (emmanuellujan) [reply src/MOLFiniteDifference/MOL_discretization.jl:203]: `r(t) !== nothing` is asking if the current rule can be applied to the current term.

Comment (ChrisRackauckas) [reply src/MOLFiniteDifference/MOL_discretization.jl:203]: It should probably just be `!=` but that's not a huge deal.

Comment (emmanuellujan) [reply src/MOLFiniteDifference/MOL_discretization.jl:203]: Great, I will fix it in the next PR :)

Comment (tinosulzer) [reply src/MOLFiniteDifference/MOL_discretization.jl:203]: that's my bad, VScode had a blue underline under `!=` which disappeared when I changed to `!==`

**Pull request merged**
